### PR TITLE
Add Listener hooks for Datasets

### LINKED ISSUE #34327: Listeners for Datasets
### Description

Add listeners for Datasets (events)

### Use case/motivation

As Airflow administrators, we would like to trigger some external processes based on all datasets being created/updated by our users. We came across the listeners for the dag runs and task instances (which are also useful), but are still missing listeners for datasets.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by yermalov-here**
This PR creates listener hooks for the following Dataset events
* on_dataset_created
* on_dataset_changed

closes: #34327 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/datasets/manager.py [hunk 1]
```diff
@@ -44,6 +45,13 @@ class DatasetManager(LoggingMixin):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
+    def create_datasets(self, dataset_models: list[DatasetModel], session: Session) -> None:
+        """Create new datasets."""
+        for dataset_model in dataset_models:
+            session.add(dataset_model)
+            self.notify_dataset_created(dataset=Dataset(uri=dataset_model.uri, extra=dataset_model.extra))
+        session.flush()  # this is required to ensure each dataset has its PK loaded
```
Comment (uranusjr) [inline airflow/datasets/manager.py]: This called the hook before the database is flushed (i.e. before the rows can actually appear in the database), while the changed hook is called after flushing. I feel this should be more consistent.

Comment (yermalov-here) [reply airflow/datasets/manager.py]: Addressed in https://github.com/apache/airflow/pull/34418/commits/c471bde88ba025d098c541b35e38d880ce368776
`on_dataset_created` hook is now called after the database is flushed.

File: airflow/models/dag.py [hunk 2]
```diff
@@ -3177,11 +3179,11 @@ def bulk_write_to_db(
                 stored_dataset.is_orphaned = expression.false()
                 stored_datasets[stored_dataset.uri] = stored_dataset
             else:
-                session.add(dataset)
-                stored_datasets[dataset.uri] = dataset
-
-        session.flush()  # this is required to ensure each dataset has its PK loaded
```
Comment (yermalov-here) [inline airflow/models/dag.py:3183]: Is it possible that this flush is not only responsible for ensuring the dataset PK availability, but for something more and that results into tests flakiness?

**Review by uranusjr — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Comment by potiuk** [comment]
Some tests shoudl be marked as DB tests- see the error and link describing what needs to be done.

**Comment by yermalov-here** [comment]
> Some tests shoudl be marked as DB tests- see the error and link describing what needs to be done.

got it. will take a look

**Comment by mobuchowski** [comment]
One question - on which Airflow component (scheduler, worker...) datasets are updated?

It would be nice to add this to docstring, probably for the rest of the methods (in some separate PR). For now it's only mentioned [here](https://github.com/apache/airflow/blob/55963de61edbbaa5f54d70f94e3f4682e824743f/docs/apache-airflow-providers-openlineage/guides/structure.rst#L26).

**Comment by yermalov-here** [comment]
> One question - on which Airflow component (scheduler, worker...) datasets are updated?
> 
> It would be nice to add this to docstring, probably for the rest of the methods (in some separate PR). For now it's only mentioned [here](https://github.com/apache/airflow/blob/55963de61edbbaa5f54d70f94e3f4682e824743f/docs/apache-airflow-providers-openlineage/guides/structure.rst#L26).

i can be wrong due to the lack of knowledge, but afaiu scheduler is responsible for creating new datasets and workers for updating them.

**Pull request merged**